### ISSUE TITLE
Interpreting the training curves on Tic Tac Toe

### ISSUE THREAD

**Issue by itmorn** [werner-duvaud/muzero-general#152]
![微信图片_20210423174341](https://user-images.githubusercontent.com/32835189/115853380-b294af80-a45b-11eb-8d46-3d535e54b33f.png)
First of all, thank you very much for opening the code.
I have trained a model with your default configuration. But the effect is not good, what may be the reason?
As far as I can tell, an agent who has mastered tic-tac-toe must have a draw on its own game, but it has decreased in episode length.

**Comment by werner-duvaud** [comment]
Hi,

What do you mean by "the effect is not good" ?

What I see is that:
* On the MuZero reward plot:
![reward](https://user-images.githubusercontent.com/40442230/115860536-ece95700-a431-11eb-889f-2a332cbb415b.png)
MuZero used to draw or lose at the start, now he wins once in 4.
The curve is quite increasing and he seems to be learning and progressing.

* On the opponnent reward:
![opponent](https://user-images.githubusercontent.com/40442230/115860572-f96daf80-a431-11eb-8606-1c656bc850e1.png)
It is clearly decreasing, he only wins once in 3 now.
In addition, it seems to me that you are testing against the expert opponent, so it is hand-coded to systematically block the winning opponent's winning opportunities.

In conclusion, with a longer training MuZero will continue to progress and the opponent will draw more and more.

**Comment by itmorn** [comment]
Thank you for your suggestion.
I trained the model for a longer time. The following figure shows the training curve:
![image](https://user-images.githubusercontent.com/32835189/116200655-2c35e180-a76b-11eb-9915-fea71f8edaf7.png)

I found out the agent is not very smart. The result of a game between me and it is as follows:
`
Welcome to MuZero! Here's a list of games:
0. atari
1. breakout
2. cartpole
3. connect4
4. gomoku
5. gridworld
6. lunarlander
7. simple_grid
8. tictactoe
9. tictactoe_gpu
10. twentyone
Enter a number to choose the game: 8
8
2021-04-27 15:30:38,877	INFO services.py:1174 -- View the Ray dashboard at http://127.0.0.1:8266

0. Train
1. Load pretrained model
2. Diagnose model
3. Render some self play games
4. Play against MuZero
5. Test the game manually
6. Hyperparameter search
7. Exit
Enter a number to choose an action: 1
1

0. results/tictactoe/2021-04-25--21-19-15/
1. results/tictactoe/2021-04-25--15-57-37/
2. results/tictactoe/2021-04-25--15-22-52/
3. results/tictactoe/2021-04-25--15-20-27/
4. results/tictactoe/2021-04-25--15-10-51/
5. results/tictactoe/2021-04-25--14-54-51/
6. results/tictactoe/2021-04-24--18-23-45/
7. results/tictactoe/2021-04-24--11-47-26/
8. results/tictactoe/2021-04-23--20-18-07/
9. Specify paths manually
Enter a number to choose a model to load: 6
6

Using checkpoint from results/tictactoe/2021-04-24--18-23-45/model.checkpoint

Initializing replay buffer with results/tictactoe/2021-04-24--18-23-45/replay_buffer.pkl

Done

0. Train
1. Load pretrained model
2. Diagnose model
3. Render some self play games
4. Play against MuZero
5. Test the game manually
6. Hyperparameter search
7. Exit
Enter a number to choose an action: 4
4
Testing 1/1
(pid=27304) [[0 0 0]
(pid=27304)  [0 0 0]
(pid=27304)  [0 0 0]]
(pid=27304) Press enter to take a step 


(pid=27304) Tree depth: 11
(pid=27304) Root value for player 0: -0.03
(pid=27304) Played action: Play row 2, column 2
(pid=27304) [[0 0 0]
(pid=27304)  [0 1 0]
(pid=27304)  [0 0 0]]
(pid=27304) Press enter to take a step 


(pid=27304) Tree depth: 8
(pid=27304) Root value for player 1: -3.62
(pid=27304) Player 1 turn. MuZero suggests Play row 1, column 1
(pid=27304) Enter the row (1, 2 or 3) to play for the player 1: 
1
1
(pid=27304) Enter the column (1, 2 or 3) to play for the player 1: 
1
1
(pid=27304) Played action: Play row 1, column 1
(pid=27304) [[ 0  0  0]
(pid=27304)  [ 0  1  0]
(pid=27304)  [-1  0  0]]
(pid=27304) Press enter to take a step 


(pid=27304) Tree depth: 8
(pid=27304) Root value for player 0: 7.18
(pid=27304) Played action: Play row 1, column 2
(pid=27304) [[ 0  0  0]
(pid=27304)  [ 0  1  0]
(pid=27304)  [-1  1  0]]
(pid=27304) Press enter to take a step 


(pid=27304) Tree depth: 9
(pid=27304) Root value for player 1: -5.44
(pid=27304) Player 1 turn. MuZero suggests Play row 3, column 2
(pid=27304) Enter the row (1, 2 or 3) to play for the player 1: 
3
3
(pid=27304) Enter the column (1, 2 or 3) to play for the player 1: 
2
2
(pid=27304) Played action: Play row 3, column 2
(pid=27304) [[ 0 -1  0]
(pid=27304)  [ 0  1  0]
(pid=27304)  [-1  1  0]]
(pid=27304) Press enter to take a step 


(pid=27304) Tree depth: 7
(pid=27304) Root value for player 0: 2.28
(pid=27304) Played action: Play row 3, column 1
(pid=27304) [[ 1 -1  0]
(pid=27304)  [ 0  1  0]
(pid=27304)  [-1  1  0]]
(pid=27304) Press enter to take a step 


(pid=27304) Tree depth: 9
(pid=27304) Root value for player 1: -4.31
(pid=27304) Player 1 turn. MuZero suggests Play row 3, column 3
(pid=27304) Enter the row (1, 2 or 3) to play for the player 1: 
3
3
(pid=27304) Enter the column (1, 2 or 3) to play for the player 1: 
3
3
(pid=27304) Played action: Play row 3, column 3
(pid=27304) [[ 1 -1 -1]
(pid=27304)  [ 0  1  0]
(pid=27304)  [-1  1  0]]
(pid=27304) Press enter to take a step 


(pid=27304) Tree depth: 9
(pid=27304) Root value for player 0: 5.61
(pid=27304) Played action: Play row 2, column 1
(pid=27304) [[ 1 -1 -1]
(pid=27304)  [ 1  1  0]
(pid=27304)  [-1  1  0]]
(pid=27304) Press enter to take a step 


(pid=27304) Tree depth: 15
(pid=27304) Root value for player 1: -19.28
(pid=27304) Player 1 turn. MuZero suggests Play row 2, column 3
(pid=27304) Enter the row (1, 2 or 3) to play for the player 1: 
2
2
(pid=27304) Enter the column (1, 2 or 3) to play for the player 1: 
3
3
(pid=27304) Played action: Play row 2, column 3
(pid=27304) [[ 1 -1 -1]
(pid=27304)  [ 1  1 -1]
(pid=27304)  [-1  1  0]]
(pid=27304) Press enter to take a step 


(pid=27304) Tree depth: 15
(pid=27304) Root value for player 0: 19.23
(pid=27304) Played action: Play row 1, column 3
(pid=27304) [[ 1 -1 -1]
(pid=27304)  [ 1  1 -1]
(pid=27304)  [-1  1  1]]
(pid=27304) Press enter to take a step 



Done

`
The unsmart action is as follows:
![image](https://user-images.githubusercontent.com/32835189/116203689-897f6200-a76e-11eb-87c3-0cc51f956c74.png)

I basically used the default configuration(self.num_workers = 7, self.selfplay_on_gpu = True, self.reanalyse_on_gpu = True). I found in the source code that the reward for both a draw and a loss is 0, so could this be the cause?

**Comment by theword** [comment]
I ran the default like you did and got the same results as you. Good graphs but Mu performs poorly in actual games. For my second experiment, I change it from "expert" to "self". And these are my results. The Loss is very good so it makes me believe the Ai learned. Both MuZero and opponent are heading to 0 reward (I believe this is expected right? More draws) but it still performs in reality poorly. 

I was also thinking the same as you. I have made a code update, Win = 1, Draw = 0 and Loss = -1. This is what DeepMind reported in their paper. I will let you know how that experiment goes. 

![image](https://user-images.githubusercontent.com/1483287/118043824-7d350080-b33b-11eb-9105-4764e44cfc0f.png)

![image](https://user-images.githubusercontent.com/1483287/118043923-9fc71980-b33b-11eb-9e9a-a880a3d96bb5.png)

**Comment by itmorn** [comment]
@theword OK, looking forward to hear from you.

**Comment by JohnPPP** [comment]
Hi,

I also tried to train it "as is" and was not able to get a good performance. As said, even though the plots show some progress, it plays quite bad.

I opened an issue with the same problem:

https://github.com/werner-duvaud/muzero-general/issues/134

Even after training for almost 10 hours I still do not reach the performance on the attached files, which I think is very good.

Hope it helps.

As anyone else had success with the training?

**Comment by theword** [comment]
> 
> 
> @theword OK, looking forward to hear from you.

Was able to get really impressive graphs but it would still lose when I played against it. Expected it to draw every time.

**Comment by JohnPPP** [comment]
> > @theword OK, looking forward to hear from you.
> 
> Was able to get really impressive graphs but it would still lose when I played against it. Expected it to draw every time.

Same here. I was not able to train it to play tic-tac-toe. Shifting to another AI algorithm since I'm unable to train anything here.

**Comment by AdrianAcala** [comment]
@JohnPPP / @theword , I've modified the tic-tac-toe a bit and got some really good results. It looks like the learning rate wasn't high enough and it should be using SGD instead of Adam. I also had to make sure there was as many workers as I had CPU threads to speed up the process. I then started tuning a learning rate decay over time. I tried it the first time, and embarrassingly, it beat me. 🤣 I've tested it and it's pretty good. I've attached a screenshot of my tensorboard showing MuZero learning strongly and opponent goes to 0.

<img width="1405" alt="image" src="https://user-images.githubusercontent.com/9313241/127112461-6069b29b-e360-487f-9cc7-f231d5edd723.png">

**Comment by JohnPPP** [comment]
@AdrianAcala , That is great! congratz on your improvements!

I've changed from this Muzero implementation to an Alpha-zero one. I must say that this one is much better designed, implemented, and easier to create new environments. However, I still cannot tune anything... It appears that the fault is on me alone and that's good.

Whishing all that come across this code great successes in the AI world,
João

**Comment by theword** [comment]
@AdrianAcala Could you share your config of the hparams?

**Comment by AdrianAcala** [comment]
@theword , I'm running into issues replicating it. The learning rate is way too high and blows up to infinity sometimes. Though, when I lower the learning rate, then it takes forever to learn.

**Comment by AdrianAcala** [comment]
@theword , AHA! I was able to increase the learning rate and not have it blow up by also increasing the batch size. I've tried this twice now and both runs show great results. Will have a PR later this evening.

**Comment by theword** [comment]
> 
> 
> @theword , AHA! I was able to increase the learning rate and not have it blow up by also increasing the batch size. I've tried this twice now and both runs show great results. Will have a PR later this evening.

Awesome. I will try it out. How many steps did it take for converging?

**Comment by ahainaut** [comment]
Hello,
Thanks for your experiments and feedbacks on hyper parameters !
Closing this issue since @AdrianAcala submitted a new set of HP (see #169).
